cabal-install/Distribution/Client/Types.hs

```diff
@@ -454,9 +466,35 @@ instance Text RelaxedDep where
           when (pkgVersion p0 == nullVersion) Parse.pfail
           pure p0
 
+instance Text RelaxDepSubject where
+  disp RelaxDepSubjectAll      = Disp.text "all"
+  disp (RelaxDepSubjectPkg pn) = disp pn
+
+  parse = (pure RelaxDepSubjectAll <* Parse.char '*') Parse.<++ pkgn
+    where
+      pkgn = do
+          pn <- parse
+          pure (if (pn == mkPackageName "all")
+                then RelaxDepSubjectAll
+                else RelaxDepSubjectPkg pn)
+
+instance Text RelaxDeps where
+  disp rd | not (isRelaxDeps rd) = Disp.text "none"
+  disp (RelaxDepsSome pkgs)      = Disp.fsep .
+                                   Disp.punctuate Disp.comma .
+                                   map disp $ pkgs
+  disp RelaxDepsAll              = Disp.text "all"
+
+  parse =           (const mempty        <$> ((Parse.string "none" Parse.+++
+                                               Parse.string "None") <* Parse.eof))
+          Parse.<++ (const RelaxDepsAll  <$> ((Parse.string "all"  Parse.+++
```

Comment:
We can, but it's not essential, as there's two ways to represent `*` currently. In a future refactoring, I do intend to make `--allow-newer` more like `--preferences` and `--constraints`, at which point the outer `RelaxDep` structure will become merely a newtype, which will avoid these corner-cases which make round-tripping tests so weird.